vms/proposervm/block/build.go

```diff
@@ -6,39 +6,140 @@ package block
+}
+
+// marshalBlock marshal the given statelessBlock by using either the default statelessBlock or
+// coping the exported fields into statelessBlockV0 and then marshaling it.
+// this allows the marsheler to produce encoded blocks that match the old style blocks as long as
+// the VRGSig feature was not enabled.
+func marshalBlock(block *statelessBlock) ([]byte, error) {
+	if len(block.StatelessBlock.VRFSig) == 0 {
+		// create a backward compatible block ( without VRFSig ) and use the statelessBlockV0 encoder for the encoding.
+		var preBlockSigBlock SignedBlock = &statelessBlockV0{
+			StatelessBlock: statelessUnsignedBlockV0{
+				ParentID:     block.StatelessBlock.ParentID,
+				Timestamp:    block.StatelessBlock.Timestamp,
+				PChainHeight: block.StatelessBlock.PChainHeight,
+				Certificate:  block.StatelessBlock.Certificate,
+				Block:        block.StatelessBlock.Block,
+			},
+			Signature: block.Signature,
+		}
+		return Codec.Marshal(CodecVersion, &preBlockSigBlock)
+	}
+	var blockIntf SignedBlock = block
+	return Codec.Marshal(CodecVersion, &blockIntf)
+}
+
+func calculateBootstrappingBlockSig(chainID ids.ID, networkID uint32) [hashing.HashLen]byte {
+	// build the hash of the following struct:
+	// +-----------------------+----------+------------+
+	// |  prefix :             | [8]byte  | "rng-root" |
+	// +-----------------------+----------+------------+
+	// |  chainID :            | [32]byte |  32 bytes  |
+	// +-----------------------+----------+------------+
+	// |  networkID:           | uint32   |  4 bytes   |
+	// +-----------------------+----------+------------+
+
+	buffer := make([]byte, 44)
+	copy(buffer, "rng-root")
+	copy(buffer[8:], chainID[:])
+	binary.LittleEndian.PutUint32(buffer[40:], networkID)
+	return hashing.Hash256(buffer)
+}
+
+func NextHashBlockSignature(parentBlockSig []byte) []byte {
+	if len(parentBlockSig) == 0 {
+		return nil
+	}
+	// previous block had a valid signature, hash that signature.
+	sigParentBlockSig := hashing.ComputeHash256(parentBlockSig)
+
+	// as long as the signature length is too short, generate additional hashes.
+	for len(sigParentBlockSig) < len(parentBlockSig) {
+		sigParentBlockSig = append(sigParentBlockSig, hashing.ComputeHash256(sigParentBlockSig)...)
+	}
+
+	// adjust the size of the hash to be as long as the parent signature ( which is a BLS signature length )
+	sigParentBlockSig = sigParentBlockSig[:len(parentBlockSig)]
+
+	return sigParentBlockSig
+}
+
+func NextBlockVRFSig(parentBlockSig []byte, blsSignKey *bls.SecretKey, chainID ids.ID, networkID uint32) []byte {
+	if blsSignKey == nil {
+		// if we need to build a block without having a BLS key, we'll be hashing the previous
+		// signature only if it presents. Otherwise, we'll keep it empty.
+		if len(parentBlockSig) == 0 {
+			// no parent block signature.
+			return []byte{}
+		}
+
+		return NextHashBlockSignature(parentBlockSig)
+	}
+
+	// we have bls key
+	var signMsg []byte
+	if parentBlockSig == nil {
```

Comment:
I don't like comparing `[]byte` types with `nil`. IMO `[]byte{}` and `[]byte(nil)` should be handled the same unless we have a very good reason not to.

Reply:
good point. I'll use `len(..)` to cover all basis.